### ISSUE TITLE
performance issues with CrossComponentIterator+AsGraphUnion+MaskSubgraph

### ISSUE THREAD

**Issue by jsichi** [jgrapht/jgrapht#757]
```
 * JGraphT version:  1.3.0
```

I instantiate AsGraphUnion with two large underlying graphs (which happen to be instances of MaskSubgraph), and then instantiate BreadthFirstIterator repeatedly on top of the union, specifying a start vertex each time.  This is very slow because

1) CrossComponentIterator eagerly instantiates entireGraphVertexIterator = graph.vertexSet().iterator(), regardless of whether start vertices are specified, and

2) AsGraphUnion.vertexSet() instantiates UnmodifiableUnionSet each time, and

3) the UnmodifiableUnionSet constructor calls size() on the underlying sets in order to decide first/second ordering, and

4) MaskSubgraph.size has to iterate over all of the underlying vertices to count wihch ones are masked

Some possible fixes:

For (1):  make the entireGraphVertexIterator instantiation lazy

For (2):  cache the UnmodifiableUnionSet (or instantiate it eagerly to avoid threading issues, although it may be preferable to keep the size checks deferred as they currently are, in case the underlying sets aren't populated yet)

Probably we should do both; @d-michail what's your opinion?

**Comment by d-michail** [comment]
Yes, we should instantiaze the `entireGraphVertexIterator` only if needed.

For (2) maybe we can skip the swap completely in the default constructor and just provide another constructor which also accepts a `Hint` enumeration regarding which set is larger.

**Comment by jsichi** [comment]
I addressed (1) in #759.  Still thinking about (2).

**Comment by jsichi** [comment]
I finally finished thinking and submitted #786.

I couldn't come up with a clean way to implement the hint suggestion, especially since it would have to be passed into AsGraphUnion as well.

Instead, I changed UnmodifiableUnionSet to defer the calls to size() until the information was actually needed.  In addition, I made it dynamic (so if the underlying set sizes change, then subsequent calls will be optimized accordingly).  This works well for my use case, but let me know what you think about it as a general solution.
